test/libsolidity/semanticTests/optimizer/overwritten_assignment_multireturn_function_variant_viaYul.sol

```diff
@@ -0,0 +1,20 @@
+// This tests that the yul optimiser
+// assigns the value of the last function
+// output to the variable on the LHS of
+// an inline assembly multi variable assignment
+// that contains the same variable more
+// than once
+contract C {
+  function f() public returns (uint o) {
+    assembly {
+      function g() -> a, b { a := 1 b := 2 }
+      let x
+      x, x := g()
+      o := x
+    }
+  }
+}
+// ====
+// compileViaYul: true
+// ----
+// f() -> 0x02
```

Comment:
Btw that that `compileViaYul` has a different evaluation order may suggest the IR generator has some other evaluation order differences we may need to check.

Reply:
Yeah, it definitely makes sense to at least investigate this a bit - the same goes for why exactly this breaks the CSE in the optimizer.

Reply:
Not sure if this PR is still necessary since #9321 makes most tests in this PR syntactically malformed.

Reply:
Probably relevant here is #7825.

Reply:
Order of evaluation seems to be consistent at first glance (between legacy and yul code gen). For example, the following test passes (one would expect different outcomes had order of evaluation been different).

```
contract C {
  function f() public returns (uint o1, uint o2) {
    assembly {
      function g() -> a, b { a := msize() mstore(0xff, 32) b := msize() }
      o1, o2 := g()
    }
  }
}
// ====
// compileViaYul: also
// ----
// f() -> 0x60, 0x120
```

Although I would have expected the first `msize()` to return `0x80` instead of `0x60` because the solidity docs says the following

> Solidity reserves four 32-byte slots, with specific byte ranges (inclusive of endpoints) being used as follows:

0x00 - 0x3f (64 bytes): scratch space for hashing methods
0x40 - 0x5f (32 bytes): currently allocated memory size (aka. free memory pointer)
0x60 - 0x7f (32 bytes): zero slot

From https://solidity.readthedocs.io/en/v0.6.5/miscellaneous.html#layout-in-memory

Reply:
Another passing example (seems to confirm consistent top-down *right-to-left* evaluation across legacy and new code gen)

```
contract C {
  function f() public returns (uint o1, uint o2) {
    assembly {
      function g() -> a, b { a := add(mload(0xff), msize()) mstore(0xff, 32) b := add(msize(), add(mload(0x120), 1)) }
      o1, o2 := g()
    }
  }
}
// ====
// compileViaYul: also
// ----
// f() -> 0x60, 0x141
```

Reply:
Right to left because the value of `a` is `0x60` which is consistent with evaluating `msize()` before `mload(0xff)` (`mload` influences `msize`); the value of `b` is `0x141` which is consistent with evaluating `mload(0x120)` *before* `msize()`

Reply:
Okay, it turns out #7825 is concerned with evaluation order for *solidity* code and not yul. Shall we close this PR @axic ?